Make FutureNCCL record events in current stream

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48790 Add support for async callbacks in ivalue::Future
* #49014 Drop FutureNCCL in favor of vanilla CUDAFuture
* #48789 Make CUDAFuture remember and restore current device in callback
* #48840 Remove DataPtr extractor from CUDAFuture
* #48788 Cache the DataPtrs in CUDAFuture
* #48506 Split out reusable CUDAFuture from FutureNCCL
* #48505 Merge common parts of FutureNCCL into at::ivalue::Future
* #48504 Split FutureNCCL's CUDA-specific parts from generic future logic
* #48502 Support wider range of types in FutureNCCL
* #48501 Don't store device indices separately on FutureNCCL
* #48500 Add multi-GPU support to FutureNCCL
* #48563 Fix FutureNCCL not recording dataptrs with caching alloc in wait()
* #48503 Fix FutureNCCL's completed() disagreeing with wait()
* #48499 Record CUDA events for "follow-up" FutureNCCL inside markCompleted
* #48498 Use fresh stream from pool for each FutureNCCL callback
* **#48497 Make FutureNCCL record events in current stream**
* #48496 Have FutureNCCL record streams w/ allocator in addCallback
* #48562 Add some safeguards to FutureNCCL
* #48495 Remove NCCL dependency from PythonFutureWrapper
* #48561 Avoid using FutureNCCL before it's ready

This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

When we record the events to mark a "follow-up" future complete (for a callback), we used to record them onto the dedicated stream, but that streams is the current stream at that time, so instead we could just record them onto the current stream. This introduces no functional differences. The reason I'm adding such an additional layer of indirection is so that the dedicated stream is only referenced inside the `addCallback` method, which will later allow us to more easily change how that stream works.

Differential Revision: [D25177553](https://our.internmc.facebook.com/intern/diff/D25177553/)

lw committed:
Make FutureNCCL record events in current stream

This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

When we record the events to mark a "follow-up" future complete (for a callback), we used to record them onto the dedicated stream, but that streams is the current stream at that time, so instead we could just record them onto the current stream. This introduces no functional differences. The reason I'm adding such an additional layer of indirection is so that the dedicated stream is only referenced inside the `addCallback` method, which will later allow us to more easily change how that stream works.

Differential Revision: [D25177553](https://our.internmc.facebook.com/intern/diff/D25177553/)

[ghstack-poisoned]

lw commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -300,8 +300,7 @@ class ProcessGroupNCCL : public ProcessGroup {
       // Use the dedicated callback stream to run callback.
-      c10::OptionalStreamGuard streamGuard{
-          c10::Stream(*futureNCCLCallbackStream_)};
+      c10::StreamGuard streamGuard{*futureNCCLCallbackStream_};
```
I also changed this so that there's a formal guarantee that we'll set a stream (i.e., remove the "Optional" part)

mrshenli commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -329,7 +328,8 @@ class ProcessGroupNCCL : public ProcessGroup {
               // records callback's stream.
-              (*thenFutCudaEvents)[0].record(*futureNCCLCallbackStream_);
+              (*thenFutCudaEvents)[0].record(
+                  at::cuda::getDefaultCUDAStream(deviceIndex_));
```
hmm, this should be current stream instead of default stream?

mrshenli replied:
Shall we also add a comment that when this cb is called, the current stream is `futureNCCLCallbackStream_`, so it's the same to record current stream here?

lw replied:
> hmm, this should be current stream instead of default stream?

Yes, absolutely! I got confused, thanks! Luckily it seems the CI also picked that up (or at least I hope that was the reason of the failures, since there shouldn't be functional changes here).

> Shall we also add a comment...

I can, if you want, but this will change very shortly (first with the current stream becoming a stream from the ATen pool, and later this being possibly any stream at all).

lw committed:
Update on "Make FutureNCCL record events in current stream"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

When we record the events to mark a "follow-up" future complete (for a callback), we used to record them onto the dedicated stream, but that streams is the current stream at that time, so instead we could just record them onto the current stream. This introduces no functional differences. The reason I'm adding such an additional layer of indirection is so that the dedicated stream is only referenced inside the `addCallback` method, which will later allow us to more easily change how that stream works.

Differential Revision: [D25177553](https://our.internmc.facebook.com/intern/diff/D25177553/)

[ghstack-poisoned]

lw committed:
Update on "Make FutureNCCL record events in current stream"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

When we record the events to mark a "follow-up" future complete (for a callback), we used to record them onto the dedicated stream, but that streams is the current stream at that time, so instead we could just record them onto the current stream. This introduces no functional differences. The reason I'm adding such an additional layer of indirection is so that the dedicated stream is only referenced inside the `addCallback` method, which will later allow us to more easily change how that stream works.

Differential Revision: [D25177553](https://our.internmc.facebook.com/intern/diff/D25177553/)

[ghstack-poisoned]

lw committed:
Update on "Make FutureNCCL record events in current stream"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

When we record the events to mark a "follow-up" future complete (for a callback), we used to record them onto the dedicated stream, but that streams is the current stream at that time, so instead we could just record them onto the current stream. This introduces no functional differences. The reason I'm adding such an additional layer of indirection is so that the dedicated stream is only referenced inside the `addCallback` method, which will later allow us to more easily change how that stream works.

Differential Revision: [D25177553](https://our.internmc.facebook.com/intern/diff/D25177553/)

[ghstack-poisoned]

mrshenli reviewed (approved):
LGTM!

lw committed:
Update on "Make FutureNCCL record events in current stream"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

When we record the events to mark a "follow-up" future complete (for a callback), we used to record them onto the dedicated stream, but that streams is the current stream at that time, so instead we could just record them onto the current stream. This introduces no functional differences. The reason I'm adding such an additional layer of indirection is so that the dedicated stream is only referenced inside the `addCallback` method, which will later allow us to more easily change how that stream works.

Differential Revision: [D25177553](https://our.internmc.facebook.com/intern/diff/D25177553/)

[ghstack-poisoned]

lw committed:
Update on "Make FutureNCCL record events in current stream"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

When we record the events to mark a "follow-up" future complete (for a callback), we used to record them onto the dedicated stream, but that streams is the current stream at that time, so instead we could just record them onto the current stream. This introduces no functional differences. The reason I'm adding such an additional layer of indirection is so that the dedicated stream is only referenced inside the `addCallback` method, which will later allow us to more easily change how that stream works.

Differential Revision: [D25177553](https://our.internmc.facebook.com/intern/diff/D25177553/)

[ghstack-poisoned]

lw committed:
Update on "Make FutureNCCL record events in current stream"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

When we record the events to mark a "follow-up" future complete (for a callback), we used to record them onto the dedicated stream, but that streams is the current stream at that time, so instead we could just record them onto the current stream. This introduces no functional differences. The reason I'm adding such an additional layer of indirection is so that the dedicated stream is only referenced inside the `addCallback` method, which will later allow us to more easily change how that stream works.

Differential Revision: [D25177553](https://our.internmc.facebook.com/intern/diff/D25177553/)

[ghstack-poisoned]

lw committed:
Update on "Make FutureNCCL record events in current stream"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

When we record the events to mark a "follow-up" future complete (for a callback), we used to record them onto the dedicated stream, but that streams is the current stream at that time, so instead we could just record them onto the current stream. This introduces no functional differences. The reason I'm adding such an additional layer of indirection is so that the dedicated stream is only referenced inside the `addCallback` method, which will later allow us to more easily change how that stream works.

Differential Revision: [D25177553](https://our.internmc.facebook.com/intern/diff/D25177553/)

[ghstack-poisoned]
